asf_lifetime_cost_model/pipeline/cost_computations.py

```diff
@@ -0,0 +1,266 @@
+
+    if subsidy_model not in supported_models:
+        raise ValueError(f"Unsupported subsidy model: {subsidy_model}. Supported models are: {supported_models}")
+
+    subsidy_data = data_getters.get_ashp_subsidy_options_data()
+    subsidy_value = subsidy_data[subsidy_data["model"] == subsidy_model][str(purchase_year)].iloc[0]
+
+    return subsidy_value
+
+
+def compute_upfront_cost(
+    heating_system: str,
+    archetype: str,
+    annual_cost_reduction: float,
+    purchase_year: int,
+    decile: int = None,
+    subsidy_model_or_input_values: Union[str, dict] = "no subsidy",
+) -> float:
+    """Computes the upfront costs of installing a heating system.
+
+    Args:
+        heating_system (str): heating system.
+            Takes "ashp" (for air source heat pump) or "boiler" (for gas boiler).
+        archetype (str): Property archetype
+            Takes: TODO: add final list of archetypes.
+        decile (int, optional): cost decile, only used for air source heat pumps.
+        annual_cost_reduction (float): how much the cost of installing the heating system reduces each year
+            in comparison to the previous year.
+        purchase_year (int): the year in which the heating system is installed.
+        subsidy_model_or_input_values (Union[str, dict], optional): The ASHP subsidy model to get the subsidy
+            to be subtracted OR a dictionary of subsidy values for each year.
+            Models include: "flat", "slow stepdown", "fast stepdown", "high", "zero from 2028",
+            "smallest", "no subsidy".
+
+    Raises:
+        ValueError: If the heating system inputed is not supported or the year is above or equal to 2035.
+
+    Returns:
+        float: Upfront cost of installing a heating system.
+    """
+    if purchase_year >= config.get("cost_year_max"):
+        raise ValueError(f"Purchase year must be before {config.get('cost_year_max')}.")
+
+    if heating_system not in ["ashp", "boiler"]:
+        raise ValueError(
+            f"Unsupported heating system: {heating_system}. Supported heating systems are `ashp` and `boiler`."
+        )
+
+    # Get installation cost for a specific heating system, archetype, and decile (where applicable)
+    installation_cost = get_installation_cost(heating_system=heating_system, archetype=archetype, decile=decile)
+
+    cost_reduction_data = create_cost_reduction_data(annual_cost_reduction=annual_cost_reduction)
+    cost_reduction_value = cost_reduction_data[purchase_year]
+
+    installation_cost = installation_cost * cost_reduction_value
+
+    if (type(subsidy_model_or_input_values) is str) and (heating_system == "ashp"):
+        subsidy_value = get_ashp_subsidy_value(subsidy_model=subsidy_model_or_input_values, purchase_year=purchase_year)
+        return installation_cost - subsidy_value
+    elif heating_system == "ashp":  # float provided
+        return installation_cost - subsidy_model_or_input_values[purchase_year]
```

Comment:
```suggestion
    elif heating_system == "ashp":  # dictionary provided
        if purchase_year not in subsidy_model_or_input_values:
            raise ValueError(f"Please provide subsidy value for purchase year: {purchase_year}")
        return installation_cost - subsidy_model_or_input_values[purchase_year]
```

Suggest adding (yet another haha) ValueError message if a subsidy value isn't provided for the purchase year.

The logic for when a dictionary is provided makes sense to me though!